bsms.mediawiki

```diff
@@ -0,0 +1,153 @@
+* The first line in the record must be the <tt>TOKEN</tt>. The second line must be the <tt>KEY</tt>. The <tt>KEY</tt> is an XPUB plus its key origin information, written in the descriptor-defined format, i.e.: <tt>[{master key fingerprint}/{derivation path}]{XPUB}</tt>. The third line must be a <tt>SIG</tt>, whereas <tt>SIG</tt> is the signature generated by using the private key associated with the XPUB to sign the first two lines.  The signature should follow [https://github.com/bitcoin/bips/blob/master/bip-0322.mediawiki BIP-0322], legacy format accepted. Finally, the Signer encrypts the entire record with <tt>ENCRYPTION_KEY</tt>.
+
+====Round 2====
+
+=====Coordinator=====
+
+* The Coordinator gathers key records from all participating Signers. Abort the setup if the wallet setup session has expired.
+* For each key record, the Coordinator decrypts it using <tt>ENCRYPTION_KEY</tt>. The Coordinator verifies that the included <tt>SIG</tt> is valid given the <tt>KEY</tt>.
+* If all key records look good, the Coordinator fills in all necessary information to generate a descriptor record, which is simply the descriptor string plus a <tt>CHECKSUM</tt>, all in one line. The <tt>CHECKSUM</tt> has [https://github.com/bitcoin/bitcoin/blob/master/doc/descriptors.md#checksums BECH32 encoding].
+* The Coordinator encrypts this descriptor record with <tt>ENCRYPTION_KEY</tt>.
+* The Coordinator sends the encrypted descriptor record to all participating Signers.
+
+=====Signer=====
+
+* The Signer imports the descriptor record, decrypts it using the <tt>ENCRYPTION_KEY</tt> derived from the open session.
+* The Signer calculates and verifies the descriptor’s <tt>CHECKSUM</tt>. Abort the setup if the <tt>CHECKSUM</tt> is incorrect.
+* The Signer checks whether one of the <tt>KEY</tt>s in the descriptor belongs to it, using path and fingerprint information included in the descriptor. The check must perform an exact match on the <tt>KEY</tt>s, and not using shortcuts such as matching fingerprints (which is trivial to spoof). Abort the setup if it doesn’t detect its own <tt>KEY</tt>.
+* For confirmation, the Signer must display to the user the <tt>CHECKSUM</tt>, plus other configurations, such as <tt>M</tt> and <tt>N</tt>. The total number of Signers, <tt>N</tt>, is important to prevent a <tt>KEY</tt> insertion attack. All participating Signers should be able to display the same confirmation.
+* If all checks pass, the Signer persists the descriptor record in its storage.
+* The Signer can choose to further restrict post-XPUB derivation paths, such as to those defined in [https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki BIP-0044].
+* The Signer should subsequently use the descriptor to generate and verify receive and change addresses.
```

Comment:
Aren't receive/change addresses usually derived at:
xpub.../0/* and xpub.../1/* ?

Reply:
bip44 have the scheme `m / purpose' / coin_type' / account' / change / address_index`. This means that xpub at 'change' level have to be derived from the hardened xkey at 'account' level. You can only have two separate xpubs for receive and change, there's no 'parent' xpub for them.

Reply:
Yes, it's the same descriptor up to the account level.
Adding this feature would help: https://github.com/bitcoin/bitcoin/issues/17190

Reply:
The difference is that the linked issue talks about defining the wallet containing two 'chains' (internal and external), when the root xpriv is present.

Here we talk about communicating xpubs, participant does not have access to other participant's xprivs. With bip44 you can only have xpubs at 'change' level, and if you want to communicate xpubs for both 'chains' to other participants, you need to send two xpubs. The current spec here only talks about one xpub per participant.

Reply:
There's actually two questions (if we allow more than one xpub per signer):

- how to ask participants to provide more than one xpub (use multiple derivation paths within one descriptor, or just multiple descriptors)
- how to communicate multiple xpubs per participant to the participant's signer device

Reply:
I don't believe there is a need for that.  That feature is designed for easier multisig setup, and many HWWs already do this (as discussed in the thread)

Reply:
Asking out loud: is "run the ceremony twice" for each chain obviously wrong? In a more general world you may want a number of chains.

Reply:
@dgpv same XPUB for change/receive addresses is how most (all?) wallets currently work. What is being shared here is the XPUB at the `account` level, if we go by BIP44.

Reply:
@dgpv also see my other comment https://github.com/nunchuk-io/bips/pull/1#discussion_r577304324

The idea is to let Signers construct/restrict their own post-XPUB subpaths (which most HWWs do already if I'm not mistaken), not that they should reuse the same path for change & receive.

Reply:
>  What is being shared here is the XPUB at the account level, if we go by BIP44.

There's no xpub at account level, because hardened derivation is used at this level.

Reply:
What do you mean? You get the XPUB for the account, then you can derive the `/0` and `/1` children from it, which are not hardened. This is how our wallet and others are doing it now.

Reply:
I was mistaken about derivation at account level, sorry for confusion. You can create xpub at account level:

```
>>> CCoinExtKey.from_seed(os.urandom(32)).derive(0x80000001).neuter()
CBitcoinExtPubKey('xpub69EvLpNRRPrNF8p43jKzULkex3pE3AxvERJDXNP7Js7ZG8z113kxhwpabMPM4brLjF6sqQuuw5VKgcJqbMDmhhXTMLFHYofiyGj1TTwiCAG')

CCoinExtKey.from_seed(os.urandom(32)).derive(0x80000001).neuter().derive(0)
CBitcoinExtPubKey('xpub6BBk4ADmkiTpmVxoH1ATADv1ZNftovChMAuHX7AtjhzDuTvLdGrUjpuufCq3fZp4bCJfzWhcVz1JGptPe9Acm7gkKFy44Vo9UEM9JMo99S9')
```